[nativert] Move MPMCQueue to torch/nativert.

zhxchen17 opened the pull request:
Summary:
Torch Native Runtime RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff adds a small library implementing a multi producer multi consumer queue which will be used to synchronize taks for Torch Native Runtime.

Differential Revision: D74184245

albanD commented on c10/util/MPMCQueue.h:
```diff
@@ -0,0 +1,61 @@
+ * A simple multi-producer, multi-consumer queue.
+ *
+ * This is a wrapper around std::deque that provides
+ * lock-free readIfNotEmpty and writeIfNotFull.
```
These functions do lock below?

zhxchen17 replied:
sorry made a typo here, will update

albanD commented on c10/test/util/MPMCQueue_test.cpp:
```diff
@@ -0,0 +1,73 @@
+#include <thread>
```
I guess new files here are automatically picked up?

zhxchen17 replied:
Yeah I took another look and seems we have a glob rule https://github.com/pytorch/pytorch/blob/main/c10/test/CMakeLists.txt

albanD commented on c10/util/MPMCQueue.h:
```diff
@@ -0,0 +1,61 @@
+ * A simple multi-producer, multi-consumer queue.
+ *
+ * This is a wrapper around std::deque that provides
+ * readIfNotEmpty and writeIfNotFull using std mutexes.
```
nit: I guess the point is for them to be thread safe. Otherwise we wouldn't need to have a custom class for this and an API designed to be easily thread-safe?
I would mention that here as a requirement in case someone changes it later. In particular, it is very important to not provide access to the queue_ directly!

albanD reviewed (approved):
Sounds good!

swolchok commented:
How many more folly things are you going to want to put into c10? I'm not excited about the idea of forking more chunks of folly and promising that we will support those forks forever in PyTorch's public API. In particular, https://github.com/facebook/folly/blob/main/folly/MPMCQueue.h is huge, presumably for good reason, so aren't we setting ourselves up to realize those reasons in the future and then have to fork and maintain the real thing?

swolchok reviewed (changes requested):
How many more folly things are you going to want to put into c10? I'm not excited about the idea of forking more chunks of folly and implicitly promising that we will support those forks forever in PyTorch's public API. In particular, https://github.com/facebook/folly/blob/main/folly/MPMCQueue.h is huge, presumably for good reason, so aren't we setting ourselves up to realize those reasons in the future and then have to fork and maintain the real thing?

zhxchen17 commented:
> How many more folly things are you going to want to put into c10? I'm not excited about the idea of forking more chunks of folly and implicitly promising that we will support those forks forever in PyTorch's public API. In particular, https://github.com/facebook/folly/blob/main/folly/MPMCQueue.h is huge, presumably for good reason, so aren't we setting ourselves up to realize those reasons in the future and then have to fork and maintain the real thing?

@swolchok Hi thanks for the comment. I don't understand what you mean by "forking folly things" in the current PR, as we don't copy any code / provide level of guarantee as folly::MPMCQueue here. Maybe we got this impression from the naming? I can change the name if it cause confusion to you.

swolchok commented:
> I don't understand what you mean by "forking folly things

As I attempted to explain, I imagine that folly::MPMCQueue is more complicated than this *for good reasons*, which implies that we are on a path to slowly rediscover those reasons and reinvent folly::MPMCQueue. Therefore, I don't expect the current simplicity of this PR to continue in the medium- to long-term.

zhxchen17 commented:
> > I don't understand what you mean by "forking folly things
> 
> As I attempted to explain, I imagine that folly::MPMCQueue is more complicated than this _for good reasons_, which implies that we are on a path to slowly rediscover those reasons and reinvent folly::MPMCQueue. Therefore, I don't expect the current simplicity of this PR to continue in the medium- to long-term.

@swolchok I think folly::MPMCQueue is complicated mostly for performance reasons because it drives RPC requests and all sorts QPS sensitive things at Meta. I don't have the intention to shift the maintenance burden to PyTorch core by adding c10::MPMCQueue here because it doesn't intend to be mimic all the observable behavior from folly::MPMCQueue. It's simply a queue that made to be linearizable across multiple consumers and producers, and perf is a none-goal here. Maybe I'm simplifying the use case here so feel free to point out what I'm missing here.

I want to understand your concern better (since I don't know what I should do instead). Are we okay with this living under torch::nativert::detail namespace and some directory folder in torch/nativert? If it's about maintenance of c10 I can place the file over there.

swolchok commented:
> Are we okay with this living under torch::nativert::detail namespace and some directory folder in torch/nativert? If it's about maintenance of c10 I can place the file over there.

I was about to suggest this idea as well after looking into what this is actually used for in nativert. (I also thought it was shimming for an internal use of folly::MPMCQueue, but it's not.) It doesn't make sense to me to promote these things to general, public-facing, PyTorch-wide utilities given their limited usage and requirements in nativert, but I would be happy to change my mind if core maintainers actively want this.

zhxchen17 committed:
[nativert] Move MPMCQueue to torch/nativert. (#152837)

Summary:

Torch Native Runtime RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff adds a small library implementing a multi producer multi consumer queue which will be used to synchronize taks for Torch Native Runtime.

Test Plan: CI

Reviewed By: dolpm

Differential Revision: D74184245

zhxchen17 commented:
> > Are we okay with this living under torch::nativert::detail namespace and some directory folder in torch/nativert? If it's about maintenance of c10 I can place the file over there.
> 
> I was about to suggest this idea as well after looking into what this is actually used for in nativert. (I also thought it was shimming for an internal use of folly::MPMCQueue, but it's not.) It doesn't make sense to me to promote these things to general, public-facing, PyTorch-wide utilities given their limited usage and requirements in nativert, but I would be happy to change my mind if core maintainers actively want this.

@swolchok yeah this makes sense. I updated the PR to put the code under torch/nativert for now.

swolchok commented on test/cpp/nativert/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ set(NATIVERT_TEST_ROOT ${TORCH_ROOT}/test/cpp/nativert)
 # Build the cpp gtest binary containing the cpp-only tests.
 set(NATIVERT_TEST_SRCS
   ${NATIVERT_TEST_ROOT}/test_tensor_meta.cpp
+  ${NATIVERT_TEST_ROOT}/test_mpmc_queue.cpp
```
nit: sort

swolchok commented on test/cpp/nativert/test_mpmc_queue.cpp:
```diff
@@ -0,0 +1,121 @@
+  std::thread reader([&queue]() {
+    for (int i = 0; i < 5; ++i) {
+      int out = 0;
+      while (!queue.readIfNotEmpty(out)) {
+        // Wait until an element is available
+      }
```
this is a bad pattern and we should not use it without great reason; you need to do something to yield the CPU in a loop like this

zhxchen17 replied:
yeah indeed, I can send a followup to address this.

swolchok reviewed (approved):
withdrawing blocking objections

yushangdi commented:
@pytorchbot merge -i 

merged internally. Need to land this to export dependent Diffs.

yushangdi commented:
@pytorchbot merge -f "merged internally. Need to land this to export dependent Diffs."

Skylion007 commented:
All C++ files should live in csrc
